numpy/core/src/umath/reduction.c

```diff
@@ -297,8 +288,50 @@ PyUFunc_ReduceWrapper(PyArrayMethod_Context *context,
         goto fail;
     }
 
+    npy_bool empty_iteration = NpyIter_GetIterSize(iter) == 0;
     result = NpyIter_GetOperandArray(iter)[0];
 
+    /*
+     * Get the initial value (if it may exists).  If the iteration is empty
+     * then we assume the reduction is (in the other case, we do not need
+     * the initial value anyway).
+     */
+    if ((initial == NULL && context->method->get_reduction_initial == NULL)
+            || initial == Py_None) {
+        /* There is no initial value, or initial value was explicitly unset */
+    }
+    else {
+        /* Not all functions will need initialization, but init always: */
+        initial_buf = PyMem_Calloc(1, op_dtypes[0]->elsize);
```

Comment:
Hmm, this makes me wonder if the initial value getter should just return a pointer to the initial value. That saves the `memcpy` too.

Reply:
Yeah, but in that set-up who free's the data?  It works beautifully if that data would be static (as is usually the case), but if we need to copy the data from something more dynamic (say an initial value for an arbitrary precision number of non-fixed size; or a structured dtype)...

Then cleanup of that data would be impossible if desired, it would have to be attached to the dtype instance or so (maybe that is OK, but not sure).